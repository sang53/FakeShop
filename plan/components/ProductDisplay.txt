inputs:
  id of product

states:
  quantity