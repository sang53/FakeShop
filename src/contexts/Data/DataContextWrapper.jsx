import { useEffect, useState } from "react";
import Helpers from "../../Helper/Helpers";
import { Products } from "./DataContext";
import PropTypes from "prop-types";

export default function DataContextWrapper({ children }) {
  const [data, setData] = useState([]);

  useEffect(() => {
    async function getData() {
      const newData = await Helpers.getProduct();
      if (useData) setData(newData);
    }
    let useData = true;
    getData();

    return () => (useData = false);
  }, []);

  return <Products.Provider value={data}>{children}</Products.Provider>;
}

DataContextWrapper.propTypes = {
  children: PropTypes.element.isRequired,
};
