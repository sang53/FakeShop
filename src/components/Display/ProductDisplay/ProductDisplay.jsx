import DisplayContainer from "../DisplayContainer/DisplayContainer";
import { use } from "react";
import { Products } from "../../../contexts/Data/DataContext";
import DisplayObj from "../DisplayObj/DisplayObj";
import { useState } from "react";
import SortSelect from "../SortSelect/SortSelect";

export default function ProductDisplay() {
  const [sort, setSort] = useState(["cat", "asc"]);
  const displayItems = use(Products).map((obj) => new DisplayObj(obj));
  sortDisplayItems(displayItems, sort);

  return (
    <>
      <h1>All Products</h1>
      <SortSelect by={sort[0]} order={sort[1]} setSort={setSort} />
      <DisplayContainer displayItems={displayItems} />
    </>
  );
}

function sortDisplayItems(displayItems, [by, order]) {
  if (by !== "cat") displayItems.sort(getSortFx(by));
  if (order === "desc") displayItems.reverse();
}

function getSortFx(by) {
  function ByAlpha(a, b) {
    const maxLength =
      a.title.length > b.title.length ? a.title.length : b.title.length;
    for (let i = 0; i < maxLength; i++) {
      const aCode = a.title[i].toLowerCase().charCodeAt();
      const bCode = b.title[i].toLowerCase().charCodeAt();
      if (aCode !== bCode) return aCode - bCode;
    }
  }

  function ByPrice(a, b) {
    return parseFloat(a.price.slice(1)) - parseFloat(b.price.slice(1));
  }

  if (by === "price") return ByPrice;
  return ByAlpha;
}
