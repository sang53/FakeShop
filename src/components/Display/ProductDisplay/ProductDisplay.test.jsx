import { render, screen } from "@testing-library/react";
import { expect, test } from "vitest";
import ProductDisplay from "./ProductDisplay";
import { Products } from "../../../contexts/Data/DataContext";
import { mockProducts } from "../../../test/utils/mockProducts";
import { MemoryRouter } from "react-router";
import { describe } from "vitest";
import { vi } from "vitest";

vi.mock("../DisplayContainer/DisplayContainer", () => ({
  default: ({ displayItems }) => {
    return (
      <div>
        {displayItems.map((displayObj) => {
          return (
            <div key={displayObj.id} data-testid="title">
              {displayObj.title}
            </div>
          );
        })}
      </div>
    );
  },
}));

function renderProducts(products = mockProducts) {
  return render(
    <MemoryRouter initialEntries={["/"]}>
      <Products.Provider value={products}>
        <ProductDisplay />
      </Products.Provider>
    </MemoryRouter>
  );
}

test("product displayed as expected", () => {
  const { container } = renderProducts(mockProducts.slice(0, 1));
  expect(container).toMatchSnapshot();
});

describe("sorts into order", () => {
  test("default order category", () => {
    renderProducts();
    expect(
      screen.getAllByTestId("title").map((element) => element.textContent)
    ).toMatchSnapshot();
  });
});
