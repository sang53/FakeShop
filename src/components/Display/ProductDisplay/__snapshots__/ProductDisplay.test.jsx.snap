// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`product displayed as expected 1`] = `
<div>
  <h1>
    All Products
  </h1>
  <div>
    <select>
      <option
        value="cat"
      >
        Category
      </option>
      <option
        value="alpha"
      >
        Alphabet
      </option>
      <option
        value="price"
      >
        Price
      </option>
    </select>
    <button>
      <svg
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
          stroke="none"
        />
        <path
          d="M12 5l0 14"
        />
        <path
          d="M18 11l-6 -6"
        />
        <path
          d="M6 11l6 -6"
        />
      </svg>
    </button>
  </div>
  <div>
    <div
      data-testid="title"
    >
      Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
    </div>
  </div>
</div>
`;

exports[`sorts into order > default order category 1`] = `
[
  "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
  "Mens Casual Premium Slim Fit T-Shirts ",
  "Mens Cotton Jacket",
  "Mens Casual Slim Fit",
  "John Hardy Women's Legends Naga Gold & Silver Dragon Station Chain Bracelet",
  "Solid Gold Petite Micropave ",
  "White Gold Plated Princess",
  "Pierced Owl Rose Gold Plated Stainless Steel Double",
  "WD 2TB Elements Portable External Hard Drive - USB 3.0 ",
  "SanDisk SSD PLUS 1TB Internal SSD - SATA III 6 Gb/s",
  "Silicon Power 256GB SSD 3D NAND A55 SLC Cache Performance Boost SATA III 2.5",
  "WD 4TB Gaming Drive Works with Playstation 4 Portable External Hard Drive",
  "Acer SB220Q bi 21.5 inches Full HD (1920 x 1080) IPS Ultra-Thin",
  "Samsung 49-Inch CHG90 144Hz Curved Gaming Monitor (LC49HG90DMNXZA) â€“ Super Ultrawide Screen QLED ",
  "BIYLACLESEN Women's 3-in-1 Snowboard Jacket Winter Coats",
  "Lock and Love Women's Removable Hooded Faux Leather Moto Biker Jacket",
  "Rain Jacket Women Windbreaker Striped Climbing Raincoats",
  "MBJ Women's Solid Short Sleeve Boat Neck V ",
  "Opna Women's Short Sleeve Moisture",
  "DANVOUY Womens T Shirt Casual Cotton Short",
]
`;
